lib/auth/predicate/builtins.go

```diff
@@ -0,0 +1,361 @@
+/*
+Copyright 2022 Gravitational, Inc.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package predicate
+
+import (
+	"reflect"
+	"regexp"
+	"strings"
+
+	"github.com/gravitational/trace"
+)
+
+func getIdentifier(obj any, selectors []string) (any, error) {
+	switch selectors[0] {
+	case "false":
+		return false, nil
+	case "true":
+		return true, nil
+	}
+
+	for _, s := range selectors {
+		if obj == nil || reflect.ValueOf(obj).IsNil() {
+			return nil, trace.BadParameter("cannot take field of nil")
+		}
+
+		if m, ok := obj.(map[string]any); ok {
```

Comment:
Just `map[string]any` or can we support e.g. `map[string][]string` too? I think we do support that for traits in other places, I'm not sure if `reflect` gives you the power to support this for any map type or if you have to enumerate them

Reply:
I don't believe it does, as noted in another comment, enumerating all possible concrete types recursively like that is tricky. I'm leaning on not doing that and generalizing to any for recursive types whenever valid. This does mean we should only create maps with an `any` typed value, through.